File: src/main/java/no/nav/veilarbveileder/service/AuthService.java
package no.nav.veilarbveileder.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.common.abac.VeilarbPep;
import no.nav.common.auth.context.AuthContextHolder;
import no.nav.common.featuretoggle.UnleashClient;
import no.nav.common.types.identer.EnhetId;
import no.nav.common.types.identer.NavIdent;
import no.nav.poao_tilgang.client.Decision;
import no.nav.poao_tilgang.client.TilgangClient;
import no.nav.veilarbveileder.client.LdapClient;
import no.nav.veilarbveileder.utils.ModiaPep;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
@Slf4j
@RequiredArgsConstructor
public class AuthService {

    private final AuthContextHolder authContextHolder;

    private final VeilarbPep veilarbPep;

    private final TilgangClient tilgangClient;

    private final ModiaPep modiaPep;

    private final LdapClient ldapClient;

    private final UnleashClient unleashClient;

    public static final String ROLLE_MODIA_ADMIN = "0000-GA-Modia_Admin";

    public NavIdent getInnloggetVeilederIdent() {
        return authContextHolder.getNavIdent().orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "NAV ident is missing"));
    }

    public String getInnloggetBrukerToken() {
        return authContextHolder.getIdTokenString().orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Token is missing"));
    }

    public boolean erSystemBruker() {
        return authContextHolder.erSystemBruker();
    }

    public void sjekkTilgangTilOppfolging() {
        if (!veilarbPep.harTilgangTilOppfolging(getInnloggetBrukerToken())) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Ikke tilgang til oppfølging");
        }
    }

    public void sjekkTilgangTilModia() {
        boolean harTilgang = modiaPep.harVeilederTilgangTilModia(getInnloggetBrukerToken());
        if (unleashClient.isEnabled("veilarbveileder.poao-tilgang.sammenligne")) {
            try {
                Decision decisionPoaoTilgang = tilgangClient.harVeilederTilgangTilModia(getInnloggetVeilederIdent().get());
                boolean harTilgangPoaoTilgang = Decision.Type.PERMIT.equals(decisionPoaoTilgang.getType());
                if (harTilgang != harTilgangPoaoTilgang) {
                    log.warn("Forskjellig resultat fra poao-tilgang og abac-modia");
                }
            } catch (Exception e) {
                log.error("Kall til poao-tilgang feilet", e);
            }
        }
        if (!harTilgang) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Ikke tilgang til modia");
        }
    }

    public void sjekkVeilederTilgangTilEnhet(EnhetId enhetId) {
        NavIdent ident = getInnloggetVeilederIdent();
        if (!harModiaAdminRolle(ident) && !veilarbPep.harVeilederTilgangTilEnhet(ident, enhetId)) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Ikke tilgang til enhet");
        }
    }

    public boolean harModiaAdminRolle(NavIdent ident) {
        return ldapClient.veilederHarRolle(ident, ROLLE_MODIA_ADMIN);
    }

}

File: src/main/java/no/nav/veilarbveileder/service/AuthService.kt
package no.nav.veilarbveileder.service

import com.nimbusds.jwt.JWTClaimsSet
import lombok.extern.slf4j.Slf4j
import no.nav.common.auth.context.AuthContextHolder
import no.nav.common.types.identer.EnhetId
import no.nav.common.types.identer.NavIdent
import no.nav.poao_tilgang.client.NavAnsattTilgangTilModiaAdminPolicyInput
import no.nav.poao_tilgang.client.NavAnsattTilgangTilModiaPolicyInput
import no.nav.poao_tilgang.client.NavAnsattTilgangTilNavEnhetPolicyInput
import no.nav.poao_tilgang.client.PoaoTilgangClient
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException
import java.text.ParseException
import java.util.*

@Service
@Slf4j
class AuthService(
    private val authContextHolder: AuthContextHolder,
    private val poaoTilgangClient: PoaoTilgangClient
) {
    val innloggetVeilederIdent: NavIdent
        get() = authContextHolder.navIdent.orElseThrow {
            ResponseStatusException(
                HttpStatus.UNAUTHORIZED,
                "NAV ident is missing"
            )
        }

    fun erSystemBruker(): Boolean {
        return authContextHolder.erSystemBruker()
    }
    fun hentInnloggetVeilederUUID(): UUID =
        authContextHolder
            .idTokenClaims.flatMap { authContextHolder.getStringClaim(it, "oid") }
            .map { UUID.fromString(it) }
            .orElseThrow { ResponseStatusException(HttpStatus.FORBIDDEN, "Fant ikke oid for innlogget veileder") }

    fun sjekkTilgangTilModia() {
        val tilgangResult = poaoTilgangClient.evaluatePolicy(NavAnsattTilgangTilModiaPolicyInput(
            hentInnloggetVeilederUUID())
        ).getOrThrow()
        if (tilgangResult.isDeny) {
            throw ResponseStatusException(HttpStatus.FORBIDDEN, "Ikke tilgang til modia")
        }
    }

    fun sjekkVeilederTilgangTilEnhet(enhetId: EnhetId?) {
        val tilgangResult = poaoTilgangClient.evaluatePolicy(
            NavAnsattTilgangTilNavEnhetPolicyInput(hentInnloggetVeilederUUID(), enhetId.toString())
        ).getOrThrow()
        if (tilgangResult.isDeny) {
            throw ResponseStatusException(HttpStatus.FORBIDDEN, "Ikke tilgang til enhet")
        }
    }

    fun harModiaAdminRolle(): Boolean {
        val tilgangResult = poaoTilgangClient.evaluatePolicy(
            NavAnsattTilgangTilModiaAdminPolicyInput(hentInnloggetVeilederUUID())
        ).getOrThrow()
        return tilgangResult.isPermit
    }

    fun erSystemBrukerFraAzureAd(): Boolean {
        return erSystemBruker() && harAADRolleForSystemTilSystemTilgang()
    }

    private fun harAADRolleForSystemTilSystemTilgang(): Boolean {
        return authContextHolder.idTokenClaims
            .flatMap { claims: JWTClaimsSet ->
                try {
                    return@flatMap Optional.ofNullable<List<String>>(claims.getStringListClaim("roles"))
                } catch (e: ParseException) {
                    return@flatMap Optional.empty<List<String>>()
                }
            }
            .orElse(emptyList())
            .contains("access_as_application")
    }

    fun erGodkjentAzureAdSystembruker(): Boolean {
        return ACCEPTLIST_AZURE_SYSTEM_USERS.contains(hentApplikasjonFraContex())
    }

    private fun hentApplikasjonFraContex(): String? {
        return authContextHolder.idTokenClaims
            .flatMap { claims: JWTClaimsSet -> getStringClaimOrEmpty(claims, "azp_name") } //  "cluster:team:app"
            .map { claim: String ->
                claim.split(":".toRegex()).dropLastWhile { it.isEmpty() }
                    .toTypedArray()
            }
            .filter { claims: Array<String> -> claims.size == 3 }
            .map { claims: Array<String> -> claims[2] }
            .orElse(null)
    }

    companion object {
        const val ROLLE_MODIA_ADMIN = "0000-GA-Modia_Admin"
        val ACCEPTLIST_AZURE_SYSTEM_USERS = listOf("veilarbfilter", "veilarbportefolje","veilarbvedtaksstotte")
        private fun getStringClaimOrEmpty(claims: JWTClaimsSet, claimName: String): Optional<String> {
            return try {
                Optional.ofNullable(claims.getStringClaim(claimName))
            } catch (e: Exception) {
                Optional.empty()
            }
        }
    }
}
